client/my-sites/post-selector/selector.jsx

```diff
@@ -97,136 +109,202 @@ const PostSelectorPosts = React.createClass( {
-			this.setState( { searchTerm: event.target.value } );
-			this.debouncedSearch();
-		}
+		return 'any' === this.props.query.type;
 	},
 
-	renderHierarchy( items, isRecursive ) {
-		const listClass = isRecursive ? 'post-selector__nested-list' : 'post-selector__list';
+	isRowLoaded( { index } ) {
+		return this.props.lastPage || !! this.getItem( index );
+	},
 
-		return (
-			<ul className={ listClass }>
-				{ items.map( this.renderItem, this ) }
-				{
-					this.props.loading && ! isRecursive ?
-					this.renderPlaceholderItem() :
-					null
-				}
-			</ul>
-		);
+	getItemHeight( item ) {
+		let height = ITEM_HEIGHT;
+
+		if ( item && item.items ) {
+			height += item.items.reduce( ( memo, nestedItem ) => {
+				return memo + this.getItemHeight( nestedItem );
+			}, 0 );
+		}
+
+		return height;
 	},
 
-	renderPlaceholderItem() {
-		const inputType = this.props.multiple ? 'checkbox' : 'radio';
+	getRowHeight( { index } ) {
+		return this.getItemHeight( this.getItem( index ) );
+	},
 
-		return (
-			<li>
-				<input className="post-selector__input" type={ inputType } name="posts" disabled={ true } />
-				<label><span className="placeholder-text">Loading list of options...</span></label>
-			</li>
-		);
+	getResultsWidth() {
+		const { selectorRef } = this.state;
+		if ( selectorRef ) {
+			return selectorRef.clientWidth;
+		}
+
+		return 0;
 	},
 
-	renderPlaceholder() {
-		return ( <ul>{ this.renderPlaceholderItem() }</ul> );
+	getRowCount() {
+		let count = 0;
+
+		if ( this.props.postsHierarchy ) {
+			count += this.props.postsHierarchy.length;
+		}
+
+		if ( ! this.props.lastPage ) {
+			count += 1;
+		}
+
+		return count;
 	},
 
 	maybeFetchNextPage() {
-		if ( this.props.lastPage || this.props.loading ) {
+		if ( this.props.loading || ! this.props.posts ) {
 			return;
 		}
 
 		this.props.onNextPage();
 	},
 
-	render() {
-		const numberPosts = this.props.posts ? this.props.posts.length : 0;
-		const showSearch = ( numberPosts > this.props.searchThreshold ) || this.state.searchTerm;
+	onSearch( event ) {
+		const searchTerm = event.target.value;
+		if ( this.state.searchTerm && ! searchTerm ) {
+			this.props.onSearch( '' );
+		}
 
-		let showTypeLabels;
-		if ( 'boolean' === typeof this.props.showTypeLabels ) {
-			showTypeLabels = this.props.showTypeLabels;
-		} else {
-			showTypeLabels = 'any' === this.props.query.type;
+		if ( searchTerm !== this.state.searchTerm ) {
+			analytics.ga.recordEvent( this.props.analyticsPrefix, 'Performed Post Search' );
+			this.setState( { searchTerm } );
+			this.debouncedSearch();
 		}
+	},
 
-		const classes = classNames(
-			'post-selector',
-			this.props.className, {
-				'is-loading': this.props.loading,
-				'is-compact': ! showSearch && ! this.props.loading,
-				'is-type-labels-visible': showTypeLabels
-			}
+	renderItem( item ) {
+		const postType = get( this.props.postTypes, [ item.type, 'labels', 'singular_name' ], '' );
```

Comment:
Although this is probably a fairly quick lookup, but would it be worth conditionally getting the postType based off of `isTypeLabelsVisible`?